### ISSUE TITLE
zhashx_insert: cannot return ENOMEM

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#5216]
While going through #5197 I was surprised to see that `zhashx_insert()`'s comments say it can only return EEXIST on error not ENOMEM.  Looking internally into the code, it appears it `assert()`s on memory errors, thus EEXIST is the only error possible.

Some code in `flux-core` assumes an error on insert is ENOMEM, should clean those cases up.  An audit might be nice too just to double check all the places its used.

Edit : I suppose technically since we built czmqcontainers inside flux-core, and we could turn off asserts, ENOMEM could be possible if there's no segfault or something (didn't dig into the code that closely) .... but I don't think that's how we want to code things, b/c perhaps someday we remove the internal czmqcontainers

**Comment by garlick** [comment]
I used to have a czmq issue (zeromq/czmq#1749) open on "fixing" czmq and had intended to do that work and submit it upstream, which is why I wanted the code in flux to assume ENOMEM was possible.  However, that work kept falling off the priority list.

**Comment by chu11** [comment]
From @garlick in #5217

> These are simple containers that do not change much upstream and that we use quite a bit. Maybe it would be OK to change our copy to not assert on allocation error?

I hadn't really thought of that, as my first thought was that we might get rid of `libczmqcontainers` someday.  It was originally added because I found bugs in the upstream implementation and we knew they wouldn't propagate to distros for some time.  So in my mind we would remove the library .... someday.

It wouldn't hurt to try and remove the asserts and return error on memory allocation instead.  Questions I think of off the top of my head:

- For some functions, like `zhash_insert()`, I think we'd have to begin setting `errno` since we wouldn't know what a `-1` return means anymore (i.e. is it EEXIST or ENOMEM).  Not sure if that could lead to cascading issues where we need/want to set errno on all functions with similar situations.  And hopefully there's no `void` returning functions that we would not to alter to return non-void (didn't look closely yet).

- A lot of the zlist/zhash functions assert on invalid input, do we want to remove asserts there as well?  And set errno = EINVAL?

- and of course we have to audit usage everywhere to see fallout.

- and maybe in `flux-sched` too since we copied `libczmqcontainers` there as well

**Comment by garlick** [comment]
I fear this is a time sink and looking at the code, possibly a lot of change.  I'm starting to feel like your original approach might be the right one (conform to the current API as published), and then if we have places where we've used these containers where we're in danger of asking for a big chunk of memory that should fail without the broker crashing (as opposed to asking for a small piece of memory and failing as collateral damage to memory pressure) then we shouldn't be using these containers.

**Comment by garlick** [comment]
> Do you have an example of the places of which you speak?

I'm not finding any.  Maybe this is no big deal

> On modern systems with overcommit, what is the probably that a reasonable memory allocation actually fails with ENOMEM?

Good point also.

To be clear: the hypothetical I was thinking of is where the broker is asked to do something unreasonable and instead of rejecting it, it crashes.  However ENOMEM handling is probably  the last line of defense in such situations anyway.

Are you ok with the general approach @chu11 is taking where we stop pretending that the czmq containers can ever return ENOMEM?  I think I am.

**Comment by chu11** [comment]
>   Are you ok with the general approach @chu11 is taking where we stop pretending that the czmq containers can ever return ENOMEM? I think I am.

Just to be clear, #5217 only handled `zhash[x]_insert()`.  Are you suggesting we expand to include everything?  `zlist_add_tail()`, `zhash_new()`, etc. etc.

**Comment by grondo** [comment]
>  I was thinking of is where the broker is asked to do something unreasonable and instead of rejecting it, it crashes

Oh yeah, if something goes wrong and someone asks for 150% of RAM, then I think malloc() could return `ENOMEM`, good point. However, it seems unlikely that could happen with the czmq stuff since it is usually allocating a small thing at a time?

> Are you ok with the general approach @chu11 is taking where we stop pretending that the czmq containers can ever return ENOMEM?

Yeah, I guess so. It seems low priority though since I don't think the inconsistency has come up ever, but I guess if someone is willing to fix it I'd be fine with it.

>  Are you suggesting we expand to include everything? zlist_add_tail(), zhash_new(), etc. etc.

I guess I'd ask if there's an issue with the `zhash[x]` functions that needs to be fixed, why would the other functions be excluded?

**Comment by chu11** [comment]
> I guess I'd ask if there's an issue with the zhash[x] functions that needs to be fixed, why would the other functions be excluded?

I guess I was thinking more about things like:

```
if (!(l = zlistx_new()))
   return NULL;
```

We've coded stuff like this **everywhere**, but internally ENOMEM is impossible b/c of the czmq asserts, so would we just stop checking for errors on this?

As a side note:

```
//  Add an item to the tail of the list. Calls the item duplicator, if any,
//  on the item. Resets cursor to list head. Returns an item handle on
//  success, NULL if memory was exhausted.
CZMQ_EXPORT void *
    zlistx_add_end (zlistx_t *self, void *item);
```

The comments say NULL if memory was exhausted, but I think that's also impossible for this case.  Maybe should do a tweak PR to upstream czmq just to double check.

**Comment by garlick** [comment]
> it seems unlikely that could happen with the czmq stuff since it is usually allocating a small thing at a time?

I think if you define an item duplicator for one of these containers and then try to insert something huge into it, failure of the duplicator triggers an assert.  So things like the content cache are maybe a concern although I haven't looked closely lately.  Anyway, it hasn't come up so perhaps not what we should obsess over right now.  Limiting message size at ingress might be the first order of business in addressing that kind of thing anyway.

> The comments say NULL if memory was exhausted, but I think that's also impossible for this case. Maybe should do a tweak PR to upstream czmq just to double check.

Yeah I just noticed that the czmq code does not seem to match the docs for zlistx.  Sigh.

**Comment by grondo** [comment]
> We've coded stuff like this everywhere, but internally ENOMEM is impossible b/c of the czmq asserts, so would we just stop checking for errors on this?

True, but it would feel so _wrong_ to not check the return value of a function that does allocation :laughing: . Also, it then introduces another inconsistency that all/most of the other creation functions _do_ need their return codes checked, but then you'd have these few cases where (you're basically requiring) they do not. In some ways that inconsistency is more dangerous than checking for errors that can't happen because of an internal assert...

However, I don't feel that strongly about it..

**Comment by chu11** [comment]
> True, but it would feel so wrong to not check the return value of a function that does allocation

My feeling is the same, it feels wrong.

When I went through my mini-audit for #5217 I did notice some places that check for errors and some that don't.  I suppose that's something we could consider cleaning up a bit consistently, that we atleast consistently check for errors or not on a few functions.

That's probably not as big a deal as what I did in #5217 though, since in those cases the wrong error would actually be returned / logged, which is absolutely what we don't want.
